File: Assignment2/cpp_source/events/event.cpp
#include <iostream>

#include "event.h"

Event ::Event(int time, int value) : EnumeratedObject("Event")
{
    this->time = time;
    this->value = value;

    std::cout << this << " created\n";
};

std::string Event::__str__()
{
    return "Event " + std::to_string(this->id) + " (T:" + std::to_string(this->time) + " V:" + std::to_string(this->value) + ")";
}

bool Event ::operator<(const Event &other)
{
    return this->time < other.time;
}

Event *Event::process()
{
    std::cout << "Processing " << this << "\n";

    Event *new_event = new Event(this->time + this->value, this->value);
    delete this;
    return new_event;
}

File: python_solution/events/event.py
"""
Module for base Event class
"""
from abc import ABC, abstractmethod

from utils import EnumeratedObject


class Event(EnumeratedObject, ABC):
    """ Base class for tracking events associated with devices. """

    def __init__(self, trigger_time, device):
        super().__init__()

        self.time = trigger_time
        self.device = device

    def __repr__(self):
        return f"{self.device} Event {self.id} T:{self.time}"

    @abstractmethod
    def process(self):
        """ Standard interface for processing of an event. """
